"Maven CI Friendly Versions" are not (fully) supported, application is incomplete/broken

**Describe the bug**
The `build` goal of the `quarkus-maven-plugin` builds an incomplete application in case ["Maven CI Friendly Versions"](https://maven.apache.org/maven-ci-friendly.html) are used.

E.g.:

Parent `pom.xml`:
```xml
    <groupId>some.project</groupId>
    <artifactId>root</artifactId>
    <version>${revision}</version>
    <packaging>pom</packaging>
    <properties>
        <!-- This is only the default for local development, CI will override it! -->
        <revision>0.1-SNAPSHOT</revision>
    </properties>
```

Submodule `pom.xml` (the one with the `quarkus-maven-plugin` execution):
```xml
  <parent>
    <groupId>some.project</groupId>
    <artifactId>root</artifactId>
    <version>${revision}</version>
  </parent>

  <artifactId>dist</artifactId>
```

As described on  ["Maven CI Friendly Versions"](https://maven.apache.org/maven-ci-friendly.html) (paragraph "Install / Deploy"), `flatten-maven-plugin` is configured properly.

**Expected behavior**
The application is built normally/fully. There should be no difference to e.g. `<version>0.1-SNAPSHOT</version>`.

**Actual behavior**
The runner jar is missing most (all?) of the application classes and the following is logged:
```
[INFO] --- quarkus-maven-plugin:1.3.0.CR2:build (default) @ dist ---
[WARNING] The POM for some.project:dist:jar:${revision} is missing, no dependency information available
[WARNING] The POM for some.project:dist:jar:${revision} is missing, no dependency information available
[INFO] [org.jboss.threads] JBoss Threads version 3.0.1.Final
[INFO] [io.quarkus.deployment.pkg.steps.JarResultBuildStep] Building thin jar: C:\Develop\some-project\dist\target\dist-0.1-SNAPSHOT-runner.jar
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Quarkus augmentation completed in 1106ms
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

**To Reproduce**
Steps to reproduce the behavior:
1. set up a small multi module project (parent + one module)
2. introduce `${revision}`
3. introduce `flatten-maven-plugin`
4. try to build the Quarkus application (e.g. `mvn clean install` on root level)

I am pretty swamped at the moment so I cannot provide a sample project right away.
But please do let me know if you need one.

_Update:_ Reproducer created, see https://github.com/quarkusio/quarkus/issues/7747#issuecomment-597336269

**Configuration**
none

**Screenshots**
none

**Environment:**
 - Output of `uname -a` or `ver`: `MINGW64_NT-10.0-18363 W4DEUMSY9003463 3.0.7-338.x86_64 2019-11-21 23:07 UTC x86_64 Msys`
 - Output of `java -version`: `OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.6+10, mixed mode)`
 - GraalVM version (if different from Java): none
 - Quarkus version or git rev: 1.3.0.CR2
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): `Apache Maven 3.6.3`

**Additional context**

I think I already found the source of the problem:
`io.quarkus.bootstrap.resolver.maven.workspace.LocalProject` is re-reading the `pom.xml` via `ModelUtils` / `MavenXpp3Reader`. This simple XML parsing does _not_ take care of the property interpolation that is required for the resolution of `${revision}`.

_One_ solution might be to prefer a possibly existing `.flattened-pom.xml` to the regular `pom.xml`.
I did something similar here: https://github.com/shrinkwrap/resolver/pull/142
Since the name and location of the flattened file is configurable, Quarkus should probably provide a config option to reflect that.

I found some time to create a reproducer, with slightly different Maven coordinates than in the original description but I think you'll manage. :-)

- clone https://github.com/famod/modmono-quarkus
- checkout `quarkus-7747-mvn-ci-friendly`

After `mvn clean install`:
- notice the size (way too small):
```
$ ll dist/target/modmono-quarkus-dist-1.0-SNAPSHOT-runner.jar
-rw-r--r-- 1 Falko 197121 83891 Mrz 10 22:45 dist/target/modmono-quarkus-dist-1.0-SNAPSHOT-runner.jar
```
- app fails to start:
```
$ java -jar dist/target/modmono-quarkus-dist-1.0-SNAPSHOT-runner.jar
Exception in thread "main" java.lang.NoClassDefFoundError: io/quarkus/runtime/Application
        at java.base/java.lang.ClassLoader.defineClass1(Native Method)
        at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1016)
        at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:174)
        at java.base/jdk.internal.loader.BuiltinClassLoader.defineClass(BuiltinClassLoader.java:800)
        at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(BuiltinClassLoader.java:698)
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:621)
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:579)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
        at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:27)
Caused by: java.lang.ClassNotFoundException: io.quarkus.runtime.Application
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
        ... 10 more
```

Now check out `master` and run `mvn clean install`:
- notice the way bigger size:
```
$ ll dist/target/modmono-quarkus-dist-1.0-SNAPSHOT-runner.jar
-rw-r--r-- 1 Falko 197121 203710 Mrz 10 22:47 dist/target/modmono-quarkus-dist-1.0-SNAPSHOT-runner.jar
```
- app starts just fine:
```
$ java -jar dist/target/modmono-quarkus-dist-1.0-SNAPSHOT-runner.jar

           ██████╗ ██╗   ██╗ █████╗ ██████╗ ██╗  ██╗██╗   ██╗███████╗    ██╗
██╗██╗    ██╔═══██╗██║   ██║██╔══██╗██╔══██╗██║ ██╔╝██║   ██║██╔════╝    ╚██╗
╚═╝╚═╝    ██║   ██║██║   ██║███████║██████╔╝█████╔╝ ██║   ██║███████╗     ╚██╗
██╗██╗    ██║▄▄ ██║██║   ██║██╔══██║██╔══██╗██╔═██╗ ██║   ██║╚════██║     ██╔╝
╚═╝╚═╝    ╚██████╔╝╚██████╔╝██║  ██║██║  ██║██║  ██╗╚██████╔╝███████║    ██╔╝
           ╚══▀▀═╝  ╚═════╝ ╚═╝  ╚═╝╚═╝  ╚═╝╚═╝  ╚═╝ ╚═════╝ ╚══════╝    ╚═╝
2020-03-10 22:47:52,289 INFO  [io.quarkus] (main) modmono-quarkus-dist 1.0-SNAPSHOT (powered by Quarkus 1.3.0.CR2) started in 1.494s. Listening on: http://127.0.0.1:8080
2020-03-10 22:47:52,320 INFO  [io.quarkus] (main) Profile prod activated.
2020-03-10 22:47:52,320 INFO  [io.quarkus] (main) Installed features: [cdi, resteasy]
```

> the dev mode does not work for your project even on master branch

Well, it seems to work for me:

**OK** on `master`:
- `mvn clean install`
- `mvn quarkus:dev -f dist/`

The same procedure _fails_ on `quarkus-7747-mvn-ci-friendly` with:
```
[...]
Caused by: javax.enterprise.inject.AmbiguousResolutionException: Ambiguous dependencies for type com.github.famod.modmono_quarkus.core.HelloBean and qualifiers [@Default]
        - java member: com.github.famod.modmono_quarkus.dist.HelloResource#bean
        - declared on CLASS bean [types=[com.github.famod.modmono_quarkus.dist.HelloResource, java.lang.Object], qualifiers=[@Default, @Any], target=com.github.famod.modmono_quarkus.dist.HelloResource]
        - available beans:
                - CLASS bean [types=[com.github.famod.modmono_quarkus.core.HelloBean, java.lang.Object], qualifiers=[@Default, @Any], target=com.github.famod.modmono_quarkus.core.HelloBean]
                - CLASS bean [types=[com.github.famod.modmono_quarkus.core.HelloBean, java.lang.Object], qualifiers=[@Default, @Any], target=com.github.famod.modmono_quarkus.core.HelloBean]
        at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:487)
        at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:420)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:224)
        ... 14 more
```

Am I missing something?
I did not touch the Quarkus version (1.3.0.CR2).

--- a/devtools/maven/src/main/java/io/quarkus/maven/BuildMojo.java
+++ b/devtools/maven/src/main/java/io/quarkus/maven/BuildMojo.java
@@ -22,6 +22,7 @@ import io.quarkus.bootstrap.app.AugmentAction;
 import io.quarkus.bootstrap.app.AugmentResult;
 import io.quarkus.bootstrap.app.CuratedApplication;
 import io.quarkus.bootstrap.app.QuarkusBootstrap;
+import io.quarkus.bootstrap.model.AppArtifact;
 import io.quarkus.bootstrap.resolver.maven.MavenArtifactResolver;
 
 /**
@@ -158,7 +159,13 @@ public class BuildMojo extends AbstractMojo {
                     .setRemoteRepositories(repos)
                     .build();
 
+            final Artifact projectArtifact = project.getArtifact();
+            final AppArtifact appArtifact = new AppArtifact(projectArtifact.getGroupId(), projectArtifact.getArtifactId(),
+                    projectArtifact.getClassifier(), projectArtifact.getArtifactHandler().getExtension(),
+                    projectArtifact.getVersion());
+
             CuratedApplication curatedApplication = QuarkusBootstrap.builder(outputDirectory.toPath())
+                    .setAppArtifact(appArtifact)
                     .setProjectRoot(project.getBasedir().toPath())
                     .setMavenArtifactResolver(resolver)
                     .setBaseClassLoader(BuildMojo.class.getClassLoader())

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/BootstrapAppModelFactory.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/BootstrapAppModelFactory.java
@@ -235,30 +235,26 @@ public class BootstrapAppModelFactory {
             return createAppModelForJar(appClasses);
         }
 
-        final LocalProject localProject = isWorkspaceDiscoveryEnabled() || enableClasspathCache
-                ? loadAppClassesWorkspace()
-                : LocalProject.load(appClasses, false);
-        LocalWorkspace workspace = null;
+        LocalProject localProject = null;
         AppArtifact appArtifact = this.appArtifact;
-        if (localProject == null) {
-            log.warn("Unable to locate maven project, falling back to classpath discovery");
-            if (appArtifact == null) {
+        if (appArtifact == null) {
+            localProject = enableClasspathCache ? loadAppClassesWorkspace() : LocalProject.load(appClasses, false);
+            if (localProject == null) {
+                log.warn("Unable to locate the maven project on the filesystem");
                 throw new BootstrapException("Failed to determine the Maven artifact associated with the application");
             }
-        } else {
-            workspace = localProject.getWorkspace();
-            if (appArtifact == null) {
-                appArtifact = localProject.getAppArtifact();
-            } else if (!appArtifact.equals(localProject.getAppArtifact())) {
-                log.warn("Provided application artifact attributes " + appArtifact +
-                        " do not match the actual project loaded from the disk " + localProject.getAppArtifact());
-            }
+            appArtifact = localProject.getAppArtifact();
         }
 
         try {
             Path cachedCpPath = null;
 
-            if (workspace != null && enableClasspathCache) {
+            LocalWorkspace workspace = null;
+            if (enableClasspathCache) {
+                if (localProject == null) {
+                    localProject = loadAppClassesWorkspace();
+                }
+                workspace = localProject.getWorkspace();
                 cachedCpPath = resolveCachedCpPath(localProject);
                 if (Files.exists(cachedCpPath)
                         && workspace.getLastModified() < Files.getLastModifiedTime(cachedCpPath).toMillis()) {

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/QuarkusBootstrap.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/QuarkusBootstrap.java
@@ -133,10 +133,7 @@ public class QuarkusBootstrap implements Serializable {
             appModelFactory.setDevMode(true);
             appModelFactory.setEnableClasspathCache(true);
         }
-        CurationResult model = appModelFactory
-                .resolveAppModel();
-        return new CuratedApplication(this, model);
-
+        return new CuratedApplication(this, appModelFactory.resolveAppModel());
     }
 
     public AppModelResolver getAppModelResolver() {

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/LocalProject.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/LocalProject.java
@@ -10,8 +10,10 @@ import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.ArrayList;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
+import java.util.Map;
 import java.util.Set;
 import org.apache.maven.model.Dependency;
 import org.apache.maven.model.Model;
@@ -28,17 +30,24 @@ public class LocalProject {
 
     private static final String PROJECT_BASEDIR = "${project.basedir}";
     private static final String POM_XML = "pom.xml";
+    private static final String REVISION = "revision";
+    static final String REVISION_EXPR = "${revision}";
 
     public static LocalProject load(Path path) throws BootstrapException {
         return load(path, true);
     }
 
     public static LocalProject load(Path path, boolean required) throws BootstrapException {
-        Path cpd = locateCurrentProjectDir(path, required);
+        final Path cpd = locateCurrentProjectDir(path, required);
         if (cpd == null) {
             return null;
         }
-        return new LocalProject(readModel(cpd.resolve(POM_XML)), null);
+        try {
+            return new LocalProject(readModel(cpd.resolve(POM_XML)), null);
+        } catch (UnresolvedRevisionException e) {
+            // if the revision couldn't be resolved, we are trying to resolve it from the workspace
+            return loadWorkspace(cpd);
+        }
     }
 
     public static LocalProject loadWorkspace(Path path) throws BootstrapException {
@@ -136,10 +145,37 @@ public class LocalProject {
         this.workspace = workspace;
         this.groupId = ModelUtils.getGroupId(rawModel);
         this.artifactId = rawModel.getArtifactId();
-        this.version = ModelUtils.getVersion(rawModel);
+
+        String version = ModelUtils.getVersion(rawModel);
+        final boolean revisionVersion;
+        if (revisionVersion = REVISION_EXPR.equals(version)) {
+            version = resolveRevision(rawModel);
+        }
+
         if (workspace != null) {
             workspace.addProject(this, rawModel.getPomFile().lastModified());
+            if (revisionVersion) {
+                if (version == null) {
+                    version = workspace.getRevision();
+                    if (version == null) {
+                        throw UnresolvedRevisionException.forGa(groupId, artifactId);
+                    }
+                } else {
+                    workspace.setRevision(version);
+                }
+            }
+        } else if (version == null) {
+            throw UnresolvedRevisionException.forGa(groupId, artifactId);
         }
+
+        this.version = version;
+    }
+
+    private static String resolveRevision(Model rawModel) {
+        final Map<String, String> props = new HashMap<>();
+        rawModel.getProperties().entrySet().forEach(e -> props.put(e.getKey().toString(), e.getValue().toString()));
+        System.getProperties().entrySet().forEach(e -> props.put(e.getKey().toString(), e.getValue().toString()));
+        return props.get(REVISION);
     }
 
     public String getGroupId() {

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/LocalWorkspace.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/LocalWorkspace.java
@@ -28,6 +28,9 @@ public class LocalWorkspace implements WorkspaceModelResolver, WorkspaceReader {
     private long lastModified;
     private int id = 1;
 
+    // value of ${revision} property
+    private String revision;
+
     protected void addProject(LocalProject project, long lastModified) {
         projects.put(project.getKey(), project);
         if (lastModified > this.lastModified) {
@@ -80,7 +83,10 @@ public class LocalWorkspace implements WorkspaceModelResolver, WorkspaceReader {
     @Override
     public File findArtifact(Artifact artifact) {
         final LocalProject lp = getProject(artifact.getGroupId(), artifact.getArtifactId());
-        if (lp == null || !lp.getVersion().equals(artifact.getVersion())) {
+        if (lp == null
+                || !lp.getVersion().equals(artifact.getVersion())
+                        && !(LocalProject.REVISION_EXPR.equals(artifact.getVersion())
+                                && lp.getVersion().equals(revision))) {
             return null;
         }
         final String type = artifact.getExtension();
@@ -112,5 +118,13 @@ public class LocalWorkspace implements WorkspaceModelResolver, WorkspaceReader {
             return Collections.emptyList();
         }
         return lastFindVersions = Collections.singletonList(artifact.getVersion());
+    }
+
+    public String getRevision() {
+        return revision;
+    }
+
+    void setRevision(String revision) {
+        this.revision = revision;
     }
 }

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/UnresolvedRevisionException.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/UnresolvedRevisionException.java
@@ -1,0 +1,17 @@
+package io.quarkus.bootstrap.resolver.maven.workspace;
+
+import io.quarkus.bootstrap.BootstrapException;
+
+public class UnresolvedRevisionException extends BootstrapException {
+
+    private static final long serialVersionUID = 1L;
+
+    public static UnresolvedRevisionException forGa(String groupId, String artifactId) {
+        return new UnresolvedRevisionException(
+                "Failed to resolve " + LocalProject.REVISION_EXPR + " for " + groupId + ":" + artifactId);
+    }
+
+    public UnresolvedRevisionException(String message) {
+        super(message);
+    }
+}

--- a/independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/workspace/test/LocalWorkspaceDiscoveryTest.java
+++ b/independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/workspace/test/LocalWorkspaceDiscoveryTest.java
@@ -11,11 +11,16 @@ import static org.junit.jupiter.api.Assertions.assertTrue;
 import io.quarkus.bootstrap.model.AppArtifactKey;
 import io.quarkus.bootstrap.resolver.maven.workspace.LocalProject;
 import io.quarkus.bootstrap.util.IoUtils;
+import java.io.File;
+import java.net.URL;
+import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.List;
 import java.util.Map;
 import org.apache.maven.model.Dependency;
 import org.apache.maven.model.Parent;
+import org.eclipse.aether.artifact.DefaultArtifact;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
@@ -216,6 +221,53 @@ public class LocalWorkspaceDiscoveryTest {
 
         assertCompleteWorkspace(project);
         assertLocalDeps(project);
+    }
+
+    @Test
+    public void testRevisionProperty() throws Exception {
+        final URL module1Url = Thread.currentThread().getContextClassLoader().getResource("workspace-revision/root/module1");
+        assertNotNull(module1Url);
+        final Path module1Dir = Paths.get(module1Url.toURI());
+        assertTrue(Files.exists(module1Dir));
+
+        final LocalProject module1 = LocalProject.load(module1Dir);
+
+        assertEquals("1.2.3", module1.getVersion());
+        assertNotNull(module1.getWorkspace()); // the revision must have been resolved from the workspace
+
+        final File root = module1.getWorkspace()
+                .findArtifact(new DefaultArtifact(module1.getGroupId(), "root", null, "pom", "${revision}"));
+        assertNotNull(root);
+        assertTrue(root.exists());
+        final URL rootPomUrl = Thread.currentThread().getContextClassLoader().getResource("workspace-revision/root/pom.xml");
+        assertEquals(new File(rootPomUrl.toURI()), root);
+    }
+
+    @Test
+    public void testRevisionPropertyOverridenWithSystemProperty() throws Exception {
+        final URL module1Url = Thread.currentThread().getContextClassLoader().getResource("workspace-revision/root/module1");
+        assertNotNull(module1Url);
+        final Path module1Dir = Paths.get(module1Url.toURI());
+        assertTrue(Files.exists(module1Dir));
+
+        final String revisionProp = "revision";
+        final String originalRevision = System.getProperty(revisionProp);
+        System.setProperty(revisionProp, "build123");
+
+        final LocalProject module1 = LocalProject.load(module1Dir);
+
+        if (originalRevision != null) {
+            System.setProperty(revisionProp, originalRevision);
+        }
+        assertEquals("build123", module1.getVersion());
+        assertNull(module1.getWorkspace()); // the workspace was not necessary to resolve the revision
+
+        final File root = LocalProject.loadWorkspace(module1Dir).getWorkspace()
+                .findArtifact(new DefaultArtifact(module1.getGroupId(), "root", null, "pom", "${revision}"));
+        assertNotNull(root);
+        assertTrue(root.exists());
+        final URL rootPomUrl = Thread.currentThread().getContextClassLoader().getResource("workspace-revision/root/pom.xml");
+        assertEquals(new File(rootPomUrl.toURI()), root);
     }
 
     private void assertCompleteWorkspace(final LocalProject project) {

--- a/independent-projects/bootstrap/core/src/test/resources/workspace-revision/root/module1/pom.xml
+++ b/independent-projects/bootstrap/core/src/test/resources/workspace-revision/root/module1/pom.xml
@@ -5,7 +5,7 @@
     <parent>
         <groupId>org.acme</groupId>
         <artifactId>root</artifactId>
-        <version>1.0</version>
+        <version>${revision}</version>
         <relativePath>../</relativePath>
     </parent>
 

--- a/independent-projects/bootstrap/core/src/test/resources/workspace-revision/root/pom.xml
+++ b/independent-projects/bootstrap/core/src/test/resources/workspace-revision/root/pom.xml
@@ -4,8 +4,12 @@
 
     <groupId>org.acme</groupId>
     <artifactId>root</artifactId>
-    <version>1.0</version>
+    <version>${revision}</version>
     <packaging>pom</packaging>
+
+    <properties>
+        <revision>1.2.3</revision>
+    </properties>
 
     <modules>
         <module>module1</module>

